### PR TITLE
Migrate block deletion E2E tests to Playwright

### PR DESCRIPTION
**Pull request by WunderBart**
## What?
Migrate block deletion E2E tests to Playwright.

## Why?
- Done as a part of the Playwright migration project: https://github.com/WordPress/gutenberg/issues/38851 (priority list),
- Should eventually resolve all [related flaky reports](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BType%5D+Flaky+Test%22+%22editor%2Fvarious%2Fblock-deletion.test.js%22+).

## How?
- Used accessible selectors where possible,
- Ensured the UI catches up when navigating with the keyboard,
- Reworded some test titles for consistency,
- Added some comments for clarity.

## Testing Instructions

The output of the following should be all ✅: 

```
npm run test:e2e:playwright -- block-deletion.spec.js
```

### CODE REVIEW

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+		).toBeFocused();
+
+		// Remove the current paragraph via the Block Toolbar options menu.
+		await editor.showBlockToolbar();
+		await editor.canvas
+			.getByRole( 'toolbar', { name: 'Block tools' } )
+			.getByRole( 'button', { name: 'Options' } )
+			.click();
+		await page
+			.getByRole( 'menuitem', { name: /Remove Paragraph/ } )
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit:
```suggestion
			.getByRole( 'menuitem', { name: /Remove Paragraph/i } )
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+			.click();
+
+		// Ensure the last block was removed.
+		await expect(
+			editor.canvas
+				.getByRole( 'document', {
+					name: 'Paragraph block',
+				} )
+				.last()
+		).toContainText( 'Second' );
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit: We can assert `editor.getBlocks` for more readable assertions:

```suggestion
		await expect.poll( editor.getBlocks ).toHaveLength( 2 );
```

Or

```suggestion
		await expect.poll( editor.getBlocks ).toMatchObject( [
			{ name: 'core/paragraph', attributes: { content: 'First' } },
			{ name: 'core/paragraph', attributes: { content: 'Second' } },
		] );
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+			name: 'core/paragraph',
+			attributes: { content: 'Second' },
+		} );
+
+		// Leave last paragraph empty and ensure it's focused.
+		await editor.insertBlock( {
+			name: 'core/paragraph',
+		} );
+		await expect(
+			editor.canvas.getByRole( 'document', { name: /Empty block/ } )
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit: 
```suggestion
			editor.canvas.getByRole( 'document', { name: /Empty block/i } )
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+		await expect(
+			editor.canvas.getByRole( 'document', {
+				name: /Empty block/i,
+			} )
+		).toBeFocused();
+
+		// Select the last two paragraphs.
+		await pageUtils.pressKeyWithModifier( 'shift', 'ArrowUp' );
+		await expect(
+			editor.canvas.locator( 'p.is-multi-selected' )
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: This is a bit unfortunate, do we have any accessible way to determine if a block is selected? Let's ditch the `p` if there's not.

```suggestion
			editor.canvas.locator( '.is-multi-selected' )
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+		// Ensure a new empty block was created and focused.
+		await expect.poll( editor.getBlocks ).toHaveLength( 3 );
+		await expect(
+			editor.canvas.getByRole( 'document', {
+				name: /Empty block/i,
+			} )
+		).toBeFocused();
+	} );
+
+	test( 'deleting all blocks', async ( { editor, page } ) => {
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js:259]: Let's keep the original title here for clarity: `results in the default block getting selected`.

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+		await page
+			.getByRole( 'menuitem', { name: /Remove Paragraph/ } )
+			.click();
+
+		// Ensure an empty block was created and focused.
+		await expect(
+			editor.canvas.getByRole( 'document', {
+				name: /Empty block/i,
+			} )
+		).toBeFocused();
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js:286]: Let's also check for the edited content after deleting all blocks:

```js
await expect.poll( editor.getBlocks ).toEqual( [] );
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+		editor,
+		page,
+	} ) => {
+		// Unregister default block type. This may happen if the editor is
+		// configured to not allow the default (paragraph) block type, either
+		// by plugin editor settings filtering or user block preferences.
+		await page.waitForFunction( () => {
+			try {
+				// eslint-disable-next-line no-undef
+				const defaultBlockName = wp.data
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Can we do `window.wp` to remove the eslint comment?

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+		// Ensure the caret is in a correct position.
+		await page.keyboard.type( '| ← caret was here' );
+		await expect(
+			editor.canvas
+				.getByRole( 'document', {
+					name: 'Paragraph block',
+				} )
+				.last()
+		).toContainText( 'Second| ← caret was here' );
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit: Same here, using `editor.getBlocks` might be easier.

**Review by kevin940726 — COMMENTED** [review]
Thank you! Really well written tests! Just left some nitpicks but overall it's good!

Note that changing the test titles means that the flaky tests reporter will no longer recognize them as the same test. If there are existing flaky test reports, we might want to close them after merging this.

File: test/e2e/specs/editor/various/block-deletion.spec.js (hunk 4 above)
Comment (WunderBart) [reply test/e2e/specs/editor/various/block-deletion.spec.js]: I couldn't really find any other indication of the selection. 😕 The `document` role elements don't support the `selected` attribute, so we can't use it in the `getByRole` locator:

```
locator.click: Error: "selected" attribute is only supported for roles: "columnheader", "gridcell", "option", "row", "rowheader", "tab", "treeitem"
```

File: test/e2e/specs/editor/various/block-deletion.spec.js (hunk 5 above)
Comment (WunderBart) [reply test/e2e/specs/editor/various/block-deletion.spec.js:259]: Not sure about this one, as it wouldn't match the format of the rest of the titles: 

```
Block deletion › deleting the last block via its options menu
Block deletion › deleting the last block via the keyboard shortcut
Block deletion › deleting the last block via backspace from an empty paragraph
Block deletion › deleting last selected block via backspace
Block deletion › deleting the last two selected blocks via backspace
Block deletion › deleting all blocks
Block deletion › deleting all blocks when the default block is unavailable
```

I assumed that adding the `...results in...` to every title would ultimately be less readable. We did that in the old `block-deletion.test.js`, where each `test` was in a separate `describe` block to (I guess) prevent the titles from reaching 300 cols 😅, e.g.:

```
block deletion - deleting the third and fourth blocks using backspace with multi-block selection results in two remaining blocks and positions the caret at the end of the second block
```

IMO, the expected result is described sufficiently with the `expect` calls. What do you think?

**Comment by WunderBart** [comment]
> Note that changing the test titles means that the flaky tests reporter will no longer recognize them as the same test. If there are existing flaky test reports, we might want to close them after merging this.

GTK, thanks @kevin940726! To be sure, do the flaky reports depend on the title only or also on the filename (since we migrate from `suite.test.js` to `suite.spec.js`)?

**Comment by WunderBart** [comment]
@kevin940726 performance specs are consistently failing with `Couldn't find "Paragraph" in the Blocks category`. Is it possible that the following bit is causing it? Are we running the performance specs against the same instance? 🤔 

https://github.com/WordPress/gutenberg/blob/aaf1276a006849d882659b439367cac300518153/test/e2e/specs/editor/various/block-deletion.spec.js#L295-L310

**Comment by kevin940726** [comment]
> To be sure, do the flaky reports depend on the title only or also on the filename?

Nope, they only depend on test title.

> Are we running the performance specs against the same instance?

It's the same WordPress instance but a different browser instance. I don't think they will impact each other, unless `removeBlockTypes` is a persistent action?

**Comment by WunderBart** [comment]
> > Are we running the performance specs against the same instance?
> 
> It's the same WordPress instance but a different browser instance. I don't think they will impact each other, unless `removeBlockTypes` is a persistent action?

Looks like it's not persistent. It also passes locally, _unfortunately_. 😞 I'll rerun and keep debugging.

**Comment by kevin940726** [comment]
No wait, it's not even the same WordPress instance on CI 😅.

**Comment by WunderBart** [comment]
> No wait, it's not even the same WordPress instance on CI 😅.

I'm seeing the same perf specs failing on other branches as well (e.g. [here](https://github.com/WordPress/gutenberg/actions/runs/4173081098) and [here](https://github.com/WordPress/gutenberg/actions/runs/4174067880)), so it's not related to the current changes. It started happening just now, so I checked what changed recently and found 2 potential candidates: #47889 and #47777. The latter introduced a check that I'd normally mark as _potentially flaky_ which might be causing the current issue. I'm trying out a fix on this PR here: https://github.com/WordPress/gutenberg/pull/48012/commits/03d5b029cbc6b1d88a5d0b747c3f3578b0ed7e7e. 

/cc @youknowriad 

edit:

Testing in 2 other PRs as well:
- https://github.com/WordPress/gutenberg/pull/48138
- https://github.com/WordPress/gutenberg/pull/48208

File: test/e2e/specs/editor/various/block-deletion.spec.js (hunk 5 above)
Comment (kevin940726) [reply test/e2e/specs/editor/various/block-deletion.spec.js:259]: I personally don't mind if the title is long or not 😅, but I see what you mean!

**Comment by youknowriad** [comment]
@WunderBart I have a similar suspicion as well. I opened this to try to debug it https://github.com/WordPress/gutenberg/pull/48094

**Comment by youknowriad** [comment]
For some reason, my fix didn't work, we can try to go with yours if it works consistently.

**Comment by youknowriad** [comment]
It looks like your proposed fix doesn't fix it either, so it must be something else.

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			editor.canvas.getByRole( 'document', {
+				name: /Empty block/i,
+			} )
+		).toBeFocused();
+
+		// Select the last two paragraphs.
+		await pageUtils.pressKeyWithModifier( 'shift', 'ArrowUp' );
+		await expect(
+			editor.canvas.locator( '.is-multi-selected' )
+		).toHaveCount( 2 );
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit: I did a similar thing in #48035 using the `getMultiSelectedBlocks` API. Maybe we can do something like that here?

```js
await expect.poll( () =>
	window.wp.data.select('core/block-editor').getMultiSelectedBlocks()
).toHaveCount( 2 );
```

File: test/e2e/specs/editor/various/block-deletion.spec.js [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+		).toBeFocused();
+
+		// Remove the current paragraph via the Block Toolbar options menu.
+		await editor.showBlockToolbar();
+		await editor.canvas
+			.getByRole( 'toolbar', { name: 'Block tools' } )
+			.getByRole( 'button', { name: 'Options' } )
+			.click();
+		await page
+			.getByRole( 'menuitem', { name: /Remove Paragraph/i } )
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-deletion.spec.js]: Nit: I think now that we're on 1.30.0, we can use string literal here.

```suggestion
			.getByRole( 'menuitem', { name: 'Remove Paragraph' } )
```

**Review by kevin940726 — APPROVED** [review]
Thanks again! A couple of suggestions but none of blocking 🙂 .

File: test/e2e/specs/editor/various/block-deletion.spec.js (hunk 9 above)
Comment (WunderBart) [reply test/e2e/specs/editor/various/block-deletion.spec.js]: I guess we're not testing the multi-block selection here, so it's OK to validate programmatically! 👍

Comment (kevin940726) [reply test/e2e/specs/editor/various/block-deletion.spec.js]: Even if we do, I don't think `.is-multi-selected` should be a public API for end users, so it should never be a testing target IMO.

Comment (WunderBart) [reply test/e2e/specs/editor/various/block-deletion.spec.js]: Could you elaborate on why you think it should not be a public API?

Comment (kevin940726) [reply test/e2e/specs/editor/various/block-deletion.spec.js]: It's a class that's never exposed to the users, nor to the plugin developers. On the other hand, [`getMultiSelectedBlocks`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-block-editor/#getmultiselectedblocks) is a documented API. While it's still not a user-facing public API, but it's better than nothing I guess. A better way might be to check for accessible attributes, like `aria-selected` or `aria-current`, but that's probably not going to be available. The next option is visual testing, but not more difficult, and should only be tested once.

**Comment by ntsekouras** [comment]
I just cherry-picked this PR to the wp/6.2 branch to get it included in the next release: caf09f17e8

**Pull request merged**